src/fstar/FStar.Interactive.Ide.Types.fst

```diff
@@ -0,0 +1,337 @@
+  | LDInterleaved (intf, impl) ->
+    Util.format2 "LDInterleaved (%s, %s)" (string_of_timed_fname intf) (string_of_timed_fname impl)
+  | LDSingle intf_or_impl ->
+    Util.format1 "LDSingle %s" (string_of_timed_fname intf_or_impl)
+  | LDInterfaceOfCurrentFile intf ->
+    Util.format1 "LDInterfaceOfCurrentFile %s" (string_of_timed_fname intf)
+  | PushFragment (Inl frag, _) ->
+    Util.format1 "PushFragment { code = %s }" frag.frag_text
+  | PushFragment (Inr d, _) ->
+    Util.format1 "PushFragment { decl = %s }" (FStar.Parser.AST.decl_to_string d)
+  | Noop -> "Noop {}"
+
+module BU = FStar.Compiler.Util
+
+let string_of_repl_stack_entry
+  : repl_stack_entry_t -> string
+  = fun ((depth, i), (task, state)) ->
+      BU.format "{depth=%s; task=%s}"
+                [string_of_int i;
+                string_of_repl_task task]
+                
+
+let string_of_repl_stack s =
+  String.concat ";\n\t\t"
+                (List.map string_of_repl_stack_entry s)
+
+let repl_state_to_string (r:repl_state)
+  : string
+  = BU.format 
+    "{\n\t\
+      repl_line=%s;\n\t\
+      repl_column=%s;\n\t\
+      repl_fname=%s;\n\t\
+      repl_cur_mod=%s;\n\t\      
+      repl_deps_stack={%s}\n\
+     }"
+     [string_of_int r.repl_line;
+      string_of_int r.repl_column;
+      r.repl_fname;
+      (match r.repl_curmod with
+       | None -> "None"
+       | Some m -> Ident.string_of_lid m.name);
+      string_of_repl_stack r.repl_deps_stack]
+
+
+let push_query_to_string pq =
+  let pk =
+    match pq.push_kind with
+    | SyntaxCheck -> "SyntaxCheck"
+    | LaxCheck -> "LaxCheck"
+    | FullCheck -> "FullCheck"
+  in
+  let code_or_decl =
+    match pq.push_code_or_decl with
+    | Inl code -> code
+    | Inr (_decl, code) -> code.code
+  in
+  FStar.Compiler.Util.format "{ push_kind = %s; push_line = %s; \
+               push_column = %s; push_peek_only = %s; push_code_or_decl = %s }"
+    [pk; string_of_int pq.push_line;
+     string_of_int pq.push_column;
+     string_of_bool pq.push_peek_only;
+     code_or_decl]
+
+let query_to_string q = match q.qq with
+| Exit -> "Exit"
+| DescribeProtocol -> "DescribeProtocol"
+| DescribeRepl -> "DescribeRepl"
+| Segment _ -> "Segment"
+| Pop -> "Pop"
+| Push pq -> "(Push " ^ push_query_to_string pq ^ ")"
+| VfsAdd _ -> "VfsAdd"
+| AutoComplete _ -> "AutoComplete"
+| Lookup _ -> "Lookup"
+| Compute _ -> "Compute"
+| Search _ -> "Search"
+| GenericError _ -> "GenericError"
+| ProtocolViolation _ -> "ProtocolViolation"
+| FullBuffer _ -> "FullBuffer"
+| Callback _ -> "Callback"
+| Format _ -> "Format"
+| Cancel _ -> "Cancel"
+
+let query_needs_current_module = function
+  | Exit | DescribeProtocol | DescribeRepl | Segment _
+  | Pop | Push { push_peek_only = false } | VfsAdd _
+  | GenericError _ | ProtocolViolation _ | FullBuffer _ | Callback _ | Format _ -> false
+  | Push _ | AutoComplete _ | Lookup _ | Compute _ | Search _ -> true
+
+let interactive_protocol_vernum = 2
```

Comment:
I explicitly did not increment the protocol version number. I only added new messages and did not want to break existing clients that may be relying on this protocol version number. I confirmed that fstar-mode.el continues to work.